File: src/haven/purus/pbot/api/PBotUtils.java
package haven.purus.pbot.api;

import haven.*;
import haven.purus.pathfinder.Pathfinder;

import java.awt.*;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Semaphore;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class PBotUtils {
	private PBotSession pBotSession;

	PBotUtils(PBotSession pBotSession) {
		this.pBotSession = pBotSession;
	}

	/**
	 * Send message to be displayed in the client
	 * @param msg Message content
	 */
	public void sysMsg(String msg) {
		pBotSession.gui.msg(msg);
	}

	/**
	 * Send message to be displayed in the client
	 * @param msg Message content
	 * @param r Red color 0-255
	 * @param g Green color 0-255
	 * @param b Blue color 0-255
	 */
	public void sysMsg(String msg, int r, int g, int b) {
		pBotSession.gui.msg(msg, new Color(r, g, b));
	}

	/**
	 * Waits until the flowermenu appears with timeout
	 * @param timeout Timeout to wait before returning null if menu is not opened
	 * @return Flowermenu that was opened
	 */
	public PBotFlowerMenu getFlowermenu(long timeout) {
		long start = System.currentTimeMillis();
		try {
			while(System.currentTimeMillis() - start < timeout) {
				FlowerMenu menu = pBotSession.gui.ui.root.getchild(FlowerMenu.class);
				if(menu != null)
					return new PBotFlowerMenu(menu);
				Thread.sleep(30);
			}
		} catch(InterruptedException ie) {}
		return null;
	}

	/**
	 * Itemact with item in hand, for example, to plant a crop or tree
	 * @param x x to click to
	 * @param y y to click to
	 * @param mod modifier for example 1 = shift etc
	 */
	public void itemact(double x, double y, int mod) {
		pBotSession.gui.map.wdgmsg("itemact", Coord.z, new Coord2d(x, y).floor(OCache.posres), mod);
	}

	/**
	 * Use to place something, for example, a stockpile
	 * @param x x place stockpile to
	 * @param y y place stockpile to
	 */
	public void placeThing(double x, double y) {
		pBotSession.gui.map.wdgmsg("place", new Coord2d(x, y).floor(OCache.posres), 0, 1, 0);
	}

	/**
	 * Click some place on map
	 * @param x x-coordinate
	 * @param y y-coordinate
	 * @param btn 1 = left click, 3 = right click
	 * @param mod 1 = shift, 2 = ctrl, 4 = alt
	 */
	public void mapClick(double x, double y, int btn, int mod) {
		pBotSession.gui.map.wdgmsg("click", getCenterScreenCoord(), new Coord2d(x, y).floor(OCache.posres), btn, mod);
	}

	/**
	 * Use to cancel stockpile placing for example
	 */
	public void cancelPlace() {
		pBotSession.gui.map.wdgmsg("place", pBotSession.gui.map.player().rc.floor(OCache.posres), 0, 3, 0);
	}

	/**
	 * Coordinates of the center of the screen
	 * @return Coordinates of the center of the screen
	 */
	public Coord getCenterScreenCoord() {
		return pBotSession.gui.map.sz.div(2);
	}

	/**
	 * Wait for the given time
	 * @param time time in milliseconds
	 */
	public static void sleep(int time) {
		try {
			Thread.sleep(time);
		} catch(InterruptedException e) {
			e.printStackTrace();
		}
	}

	/**
	 * Returns the item currently in the hand
	 * @return Item at hand
	 */
	public PBotItem getItemAtHand() {
		if(pBotSession.gui.vhand == null)
			return null;
		else
			return new PBotItem(pBotSession.gui.vhand.item, pBotSession);
	}

	/**
	 * Left click to somewhere with pathfinder
	 * @param x X-Coordinate
	 * @param y Y-Coordinate
	 */
	public void pfLeftClick(double x, double y) {
		Pathfinder.run(new Coord2d(x, y),null, 1, 0, -1,"", pBotSession.gui);
	}

	/**
	 * Waits for the hourglass timer when crafting or drinking for example
	 * Also waits until the hourglass has been seen to change at least once
	 * If hourglass does not appear within timeout, returns false, else true
	 * @param timeout Timeout in milliseconds
	 */
	public boolean waitForHourglass(int timeout) {
		double prog = pBotSession.gui.prog;
		int retries = 0;
		while(prog == pBotSession.gui.prog) {
			if(retries > timeout / 5)
				return false;
			retries++;
			sleep(5);
		}
		while(pBotSession.gui.prog >= 0) {
			sleep(25);
		}
		return true;
	}

	/**
	 * Starts crafting item with the given name
	 * @param name Name of the item ie. "clogs"
	 * @param makeAll 0 To craft once, 1 to craft all
	 */
	public void craftItem(String name, int makeAll) {
		openCraftingWnd(name);
		pBotSession.gui.makewnd.makeWidget.wdgmsg("make", makeAll);
	}

	/**
	 * Opens the crafting window for given item
	 * @param name Name of craft for wdgmsg
	 */
	public void openCraftingWnd(String name) {
		// Close current window and wait for it to close
		if(pBotSession.gui.makewnd.makeWidget != null)
			pBotSession.gui.makewnd.makeWidget.reqdestroy();
		pBotSession.gui.wdgmsg("act", "craft", name);
		pBotSession.PBotWindowAPI().waitForWindow("Crafting", 60*1000);
		while(pBotSession.gui.makewnd.makeWidget == null) {
			PBotUtils.sleep(10);
		}
	}

	/**
	 * Waits for the hourglass timer when crafting or drinking for example
	 * Also waits until the hourglass has been seen to change at least once
	 */
	public void waitForHourglass() {
		double prog = pBotSession.gui.prog;
		while (prog == pBotSession.gui.prog) {
			prog = pBotSession.gui.prog;
			sleep(5);
		}
		while (pBotSession.gui.prog >= 0) {
			sleep(20);
		}
	}

	/**
	 * Returns value of hourglass, -1 = no hourglass, else the value between 0.0 and 1.0
	 * @return value of hourglass
	 */
	public double getHourglass() {
		return pBotSession.gui.prog;
	}

	/**
	 * Returns the players inventory
	 * @return Inventory of the player
	 */
	public PBotInventory playerInventory() {
		return new PBotInventory(pBotSession.gui.maininv, pBotSession);
	}

	/**
	 * Drops an item from the hand and waits until it has been dropped
	 * @param mod 1 = shift, 2 = ctrl, 4 = alt
	 */
	public void dropItemFromHand(int mod, boolean wait) {
		pBotSession.gui.map.wdgmsg("drop", Coord.z, pBotSession.gui.map.player().rc.floor(OCache.posres), mod);
		if(wait) {
			while(getItemAtHand() != null)
				sleep(25);
		}
	}

	/**
	 * Resource name of the tile in the given location
	 * @param x X-Coord of the location (rc coord)
	 * @param y Y-Coord of the location (rc coord)
	 * @return
	 */
	public String tileResnameAt(int x, int y) {
		while(true) {
			try {
				Coord loc = new Coord(x, y);
				int t = pBotSession.gui.ui.sess.glob.map.gettile(loc.div(MCache.tilesz2));
				Resource res = pBotSession.gui.ui.sess.glob.map.tilesetr(t);
				if(res != null)
					return res.name;
				else
					return null;
			} catch(Loading l) {
				Thread.onSpinWait();
			}
		}
	}

	/**
	 * Next click to item in inventory calls the callback with PBotItem object of the clicked item
	 */
	public void selectItem(Callback cb) {
		pBotSession.PBotUtils().sysMsg("Left click item in inventory to select it!");
		synchronized(pBotSession.gui.itemCallbacks) {
			pBotSession.gui.itemCallbacks.add(new Pair<>(cb, pBotSession));
		}
	}

	public static class AreaReturn {
		private Coord a;
		private Coord b;
		public AreaReturn(Coord a, Coord b) {
			this.a = a.min(b.x, b.y);
			this.b = a.max(b.x, b.y);
		}

		public Coord getA() {
			return a;
		}

		public Coord getB() {
			return b;
		}
	}

	/**
	 * Select area returns AreaReturn object
	 * @param cb
	 */
	public void selectArea(Callback cb) {
		sysMsg("Select area by left clicking and dragging ground!");
		synchronized(pBotSession.gui.map.areaSelectCbQueue) {
			pBotSession.gui.map.areaSelectCbQueue.add(new Pair<>(cb, pBotSession));
		}
	}

	// RC coords for this session, may return null if grid id not found
	public Coord2d getCoords(long gridId, double ofsX, double ofsY, boolean wait) {
		while(true) {
			MCache.Grid g = pBotSession.gui.ui.sess.glob.map.getgrid(gridId);
			if(g == null)
				if(!wait)
					return null;
				else {
					PBotUtils.sleep(50);
				}
			else
				return g.gc.mul(MCache.cmaps).mul(MCache.tilesz).add(ofsX, ofsY);
		}
	}

	public void give() {
		if(pBotSession.gui.fv.current != null) {
			pBotSession.gui.fv.wdgmsg("give", (int) pBotSession.gui.fv.current.gobid, 1);
		}
	}

	public int combatState() {
		if(pBotSession.gui.fv.current != null) {
			return pBotSession.gui.fv.curgive.state;
		}
		return -1;
	}

	// Wait for the pathfinder to stop, returns true if route finding was successful false otherwise
	// Timeout in milliseconds
	public boolean pfWait(int timeout) {
		try {
			if(pBotSession.gui.map.pf_route_found == null)
				return false;
			return pBotSession.gui.map.pf_route_found.get(timeout, TimeUnit.MILLISECONDS);
		} catch(TimeoutException e) {
		} catch(InterruptedException | ExecutionException ie) {
			ie.printStackTrace();
		}
		return false;
	}

	public boolean hasCombat() {
		return pBotSession.gui.fv.current != null;
	}

	public double combatCooldown() {
		return Math.max(0.0, pBotSession.gui.fv.atkct - Utils.rtime());
	}
}

File: scripts/__pbot/PBotUtils.py
from .PBotFlowerMenu import PBotFlowerMenu
from .PBotItem import PBotItem
from .PBotInventory import PBotInventory


from . import PBotSession
from typing import Optional, Tuple, Callable

class PBotUtils(object):
    def __init__(self, session):
        self.__session = session

    ## Send a colored message in the client system menu
    # @param msg: str Message to send
    # @param r: red color between 0-255, default 255
    # @param g: green color between 0-255, default 255
    # @param b: blue color between 0-255, default 255
    def sys_msg(self, msg: str, r: int = 255, g: int = 255, b: int = 255):
        self.__session.PBotUtils().sysMsg(msg, r, g, b)

    ## Waits until the flowermenu appears with timeout
    # @param timeout timeout to wait before returning null if menu is not opened, in milliseconds
    # @return Flowermenu or None if not found within the timeout
    def get_flowermenu(self, timeout: int = 3600000) -> Optional[PBotFlowerMenu]:
        menu = self.__session.PBotUtils().getFlowermenu(timeout)
        return PBotFlowerMenu(menu) if menu is not None else None

    ## Itemact with item in hand, for example to make a stockpile
    def make_pile(self):
        self.__session.PBotUtils().itemact(0.0, 0.0, 0)

    ## Itemact with item in hand, for example, to plant a crop or a tree
    # @param x x to click to
    # @param y y to click to
    # @param mod modifier for example 1 = shift etc
    def itemact(self, x: float, y: float, mod: int = 0):
        self.__session.PBotUtils().itemact(x, y, mod)

    ## Use to place something, for example a stockpile
    # @param x x to place stockpile to
    # @param y y to place stockpile to
    def placeThing(self, x: float, y: float):
        self.__session.PBotUtils().placeThing(x, y)

    ## Click some place on map
    # @param x x to click
    # @param y y to click
    # @param btn 1 = left click, 3 = right click
    # @param mod 1 = shift, 2 = ctrl, 4 = alt
    def map_click(self, x: float, y: float, btn: int = 1, mod: int = 0):
        self.__session.PBotUtils().mapClick(x, y, btn, mod)

    ## Use to cancel stockpile placing for example
    def cancel_place(self):
        self.__session.PBotUtils().cancelPlace()

    ## Coordinates of the center of the screen
    # @return Coordinates of the sceen center
    def get_center_screen_coord(self) -> Tuple[int, int]:
        c = self.__session.PBotUtils().getScenterScreenCoord()
        return c.x, c.y,

    ## Returns the item currently at hand
    # @return item at hand or None if not found
    def get_item_at_hand(self) -> Optional[PBotItem]:
        item = self.__session.PBotUtils().getItemAtHand()
        return PBotItem(item) if item is not None else None

    ## Left click somewhere with pathfinder
    # @param x X-Coordinate
    # @param y Y-Coordinate
    def pf_left_click(self, x: float, y: float):
        self.__session.PBotUtils().pfLeftClick(x, y)

    ## Waits for the hourglass timer when crafting or drinking for example
    # Also waits until the hourglass has been seen to change at least once
    # @param timeout timeout in milliseconds
    # @return If hourglass does not appear within timeout, returns false, else true
    def wait_for_hourglass(self, timeout: int = 3600000) -> bool:
        return self.__session.PBotUtils().waitForHourglass(timeout)

    ## Starts crafting item with the given name
    # @param name name of the item ie. "clogs"
    # @param make_all 0 to craft once, 1 to craft all
    def craft_item(self, name: str, make_all: int = 0):
        self.__session.PBotUtils().craftItem(name, make_all)

    ## Returns value of the hourglass
    # @return -1 = no hourglass else the value is between 0.0 and 1.0
    def get_hourglass(self) -> float:
        return self.__session.PBotUtils().getHourglass()

    ## Returns the players inventory
    # @return Inventory of the player
    def player_inventory(self) -> PBotInventory:
        return PBotInventory(self.__session.PBotUtils().playerInventory())

    ## Drops an item from the hand and optionally waits until it has been dropped
    # @param mod 1 = shift, 2 = ctrl, 4 = alt
    # @param wait if true, wait until the item has been dropped
    def drop_item_from_hand(self, mod: int = 0, wait: bool = False):
        self.__session.PBotUtils().dropItemFromHand(mod, wait)

    ## Resource name of the tile in the given location
    # @param x X-coord of the location (rc-coord)
    # @param y Y-coord of the location (rc-coord)
    # @return tile resname or None if couldnt determine
    def tile_resname_at(self, x: int, y: int) -> Optional[str]:
        return self.__session.PBotUtils().tileResnameAt(x, y)

    ## Next click to item in inventory calls the callback with PBotItem object of the clicked item
    # @param cb callback called with the item
    def select_item(self, cb: Callable[[PBotItem], any]):
        self.__session.PBotUtils().selectItem(_SelectItemCb(cb))

    ## Select area by dragging
    # @param cb callback called with the selected area
    def select_area(self, cb: Callable[[Tuple[float, float], Tuple[float, float]], any]):
        self.__session.PBotUtils().selectArea(_SelectAreaCb(cb))

    ## Wait for pathfinder to finish what its doing
    # @param timeout timeout in milliseconds before returning false if not finished
    # @return true if route was found and executed, false if not
    def pf_wait(self, timeout: int = 999999):
        return self.__session.PBotUtils().pfWait(timeout)

    ## Get rc coords of some gridid offset pair
    # @param grid_id mapgrid id
    # @param ofs_x x offset in mapgrid
    # @param ofs_y y offset in mapgrid
    # @param wait until the mapgrid has been loaded
    # @return coords of None if grid couldn't be found
    def get_coords(self, grid_id: int, ofs_x: float, ofs_y: float, wait: bool = True) -> Optional[Tuple[float, float]]:
        c = self.__session.PBotUtils().getCoords(grid_id, float(ofs_x), float(ofs_y), wait)
        if c is None:
            return None
        else:
            return (c.x, c.y)


class _SelectItemCb(object):
    def __init__(self, cb: Callable[[PBotItem], any]):
        self.cb = cb

    def callback(self, itm):
        self.cb(PBotItem(itm))

    class Java:
        implements = ["haven.purus.pbot.api.Callback"]



class _SelectAreaCb(object):
    def __init__(self, cb: Callable[[Tuple[float, float], Tuple[float, float]], any]):
        self.cb = cb

    def callback(self, area):
        self.cb((float(area.getA().x), float(area.getA().y),), (float(area.getB().x), float(area.getB().y),))

    class Java:
        implements = ["haven.purus.pbot.api.Callback"]